### PR TITLE
Improve ResourceIndexer tests

### PR DESCRIPTION
**Pull request by epicadk**
Fixes: #282

**Description:**
Clear and concise code change description. 

1.  Renamed unit tests to signify the code path they test. 
2.  Use minimally constructed resources to test specific types
3. Add integration tests that use resources parsed from JSON files and verify that all indices are created.
4. Removed setup method so that only the resource being used in the test is parsed as compared to all resources when using the setup method. 

**Type** 
  Testing

**Screenshots (if applicable)**

### CODE REVIEW

**Comment by jingtang10** [comment]
is there any update on this one? thanks!

**Comment by jingtang10** [comment]
please make sure @deepankarb also has a chance to review once done. thanks!

**Comment by epicadk** [comment]
> is there any update on this one? thanks!

Sorry for the delay. Will be done within a day.

**Comment by epicadk** [comment]
Sorry something urgent came up. I've completed adding unit tests. Will finish integration tests soon.

**Comment by epicadk** [comment]
I need to place the integration tests in a different folder. I was thinking I can create a separate directory for unit tests and integration tests under the test directory. Any suggestions?

**Comment by epicadk** [comment]
I've also added some tests in #332 for the date types. I've covered all cases except the integer case of the number index. I can't seem to find a resource that uses a search parameter of that type.  Even the example given on the specification website doesn't work.

File: core/src/test/java/com/google/android/fhir/ResourceIndexerTest.kt [hunk 1]
```diff
@@ -0,0 +1,280 @@
+/*
```
Comment (jingtang10) [inline core/src/test/java/com/google/android/fhir/ResourceIndexerTest.kt:1]: I don't think it's right to create another file in this package. This should be in the same directory as the other file.

**Review by jingtang10 — CHANGES_REQUESTED** [review]
It might be ok to keep everything in the same file, but you should have two classes of test cases

1. cases that test specific code paths e.g. `string_empty_shouldIndexNothing` or `int_shouldCreateIndex` etc etc. We would use minimally constructed resources to test specific types.
2. cases where we use more complex resources and assert that all indecies that should be created are created.

I think these two classes of tests can probably co-exist in the same file if named sensibly.

File: core/src/test/java/com/google/android/fhir/ResourceIndexerTest.kt (hunk 1 above)
Comment (epicadk) [reply core/src/test/java/com/google/android/fhir/ResourceIndexerTest.kt]: > I don't think it's right to create another file in this package. This should be in the same directory as the other file.

Reverted this change

**Comment by epicadk** [comment]
I've kept the order similar as they occur in the resource indexer. I've also included tests for null and empty values wherever possible.

File: engine/src/test/java/com/google/android/fhir/index/ResourceIndexerTest.kt [hunk 2]
```diff
@@ -26,548 +26,803 @@ import com.google.android.fhir.index.entities.ReferenceIndex
...
-        MolecularSequence::class.java,
-        "/number_test_molecular_sequence.json"
-      )
-    specialTestLocation =
-      testingUtils.readFromFile(Location::class.java, "/location-example-hl7hq.json")
+    testPatient = testingUtils.readFromFile(Patient::class.java, "/date_test_patient.json")
+    testLocation = testingUtils.readFromFile(Location::class.java, "/location-example-hl7hq.json")
   }
 
+  /** Integration tests for ResourceIndexer. */
```
Comment (jingtang10) [inline engine/src/test/java/com/google/android/fhir/index/ResourceIndexerTest.kt:87]: can you move the 'integration' tests after the 'unit' tests?

File: engine/src/test/java/com/google/android/fhir/index/ResourceIndexerTest.kt [hunk 3]
```diff
@@ -26,548 +26,803 @@ import com.google.android.fhir.index.entities.ReferenceIndex
...
 
+  /** Unit tests for resource indexer */
   @Test
-  fun index_patient_emptyGivenName_shouldNotIndexGivenName() {
-    val resourceIndices = ResourceIndexer.index(TEST_PATIENT_NULL_FIELDS)
-    assertThat(
-        resourceIndices.stringIndices.any { stringIndex ->
-          stringIndex.path.equals("Patient.name.given")
-        }
+  fun index_Patient_shouldIndexLastUpdatedDate() {
```
Comment (jingtang10) [inline engine/src/test/java/com/google/android/fhir/index/ResourceIndexerTest.kt:322]: the point of these unit test cases is to cover all code paths so i think they need to be re-organized by index type, and please rename them, e.g.:

```
fun index_string_null
fun index_string
...

fun date_period_someCondition
fun date_period_someOtherCondition
...
```

**Review by jingtang10 — CHANGES_REQUESTED** [review]
A couple of high-level comments. Please address! Thanks!

**Comment by epicadk** [comment]
Noticed a bug in the date indexer while testing so I made the corresponding change. The value to should be inclusive (makes querying easier using the SQL between operator.)

**Comment by jingtang10** [comment]
@epicadk did you remove all the other changes in the test class you made?

**Comment by epicadk** [comment]
> @epicadk did you remove all the other changes in the test class you made?

Yup all of the tests are in the same file and I've renamed the tests as requested [here](https://github.com/google/android-fhir/pull/387#discussion_r616014331). I've also updated the pr description to explain the changes I have made.

File: engine/src/test/java/com/google/android/fhir/index/ResourceIndexerTest.kt [hunk 4]
```diff
@@ -26,542 +26,831 @@ import com.google.android.fhir.index.entities.ReferenceIndex
...
+          "Invoice.participant.role",
+          testInvoice.participantFirstRep.role.codingFirstRep.system,
+          testInvoice.participantFirstRep.role.codingFirstRep.code
+        )
+      )
+
+    assertThat(resourceIndices.uriIndices).isEmpty()
+
+    assertThat(resourceIndices.dateIndices)
       .contains(
```
Comment (jingtang10) [inline engine/src/test/java/com/google/android/fhir/index/ResourceIndexerTest.kt:657]:
```suggestion
      .containsExactly(
```

**Review by jingtang10 — CHANGES_REQUESTED** [review]
@epicadk Really happy with this - this is a very good PR! Thank you so much for this!

Apologies for the delay in reviewing. Please let me know once you're done with the few small comments I made.

**Comment by epicadk** [comment]
@jingtang10 divided the tests in 3 parts.

1.  The creation of resources 
2. Actual code for indexing
3. assertions

**Review by jingtang10 — APPROVED** [review]
Great! Thanks!
